dbms/src/DataTypes/JSONB/JSONBSerialization.cpp

```diff
@@ -0,0 +1,404 @@
+#include <Columns/ColumnString.h>
+#include <Columns/JSONB/JSONBDataMark.h>
+#include <DataTypes/DataTypeString.h>
+#include <DataTypes/JSONB/JSONBStreamBuffer.h>
+#include <DataTypes/JSONB/JSONBStreamFactory.h>
+
+#if !__clang__
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wattributes"
+#pragma GCC diagnostic ignored "-Wunused-parameter"
+#endif
+#include <Fleece/Encoder.hh>
+#if !__clang__
+#pragma GCC diagnostic pop
+#endif
+
+namespace DB
+{
+namespace ErrorCodes
+{
+    extern const int CANNOT_PARSE_JSON;
+}
+
+namespace
+{
+    const static DataTypePtr & BINARY_JSON_TYPE = std::make_shared<DataTypeString>();
+    const static DataTypePtr & NUMBER_DATA_TYPE = std::make_shared<DataTypeUInt64>();
+    const static DataTypePtr & STRING_DATA_TYPE = std::make_shared<DataTypeString>();
+    const static DataTypePtr & BOOLEAN_DATA_TYPE = std::make_shared<DataTypeUInt8>();
+
+//    inline UInt8 maybeFillingJSONTypeToMark(const ColumnJSONBStructPtr & struct_info)
+//    {
+//        return struct_info->data_columns.size() == 1 && !struct_info->children.empty() ? UInt8(TypeIndex::JSONB) : UInt8(TypeIndex::Nothing);
+//    }
+//
+//    static inline void normalizeForDeserialize(ColumnJSONB & column, size_t old_column_size)
+//    {
+//        std::vector<ColumnJSONB::WrappedPtr> & marks = column.getMarks();
+//        std::vector<ColumnJSONB::WrappedPtr> & fields = column.getFields();
+//
+//        for (size_t index = 0; index < marks.size(); ++index)
+//            if (marks[index]->size() == old_column_size)
+//                marks[index]->insertDefault();
+//
+//        for (size_t index = 0; index < fields.size(); ++index)
+//            if (fields[index]->size() == old_column_size)
+//                fields[index]->insertDefault();
+//    }
+    static inline void storeBinaryJSON(IColumn & column_, const fleece::slice & binary)
+    {
+        ColumnJSONB & column = static_cast<ColumnJSONB &>(column_);
+        const auto & column_struct = column.getStruct();
+        auto & binary_children_column = static_cast<ColumnString &>(*column_struct->getOrCreateDataColumn(BINARY_JSON_TYPE));
+
+        ColumnString::Chars & data = binary_children_column.getChars();
+        ColumnString::Offsets & offsets = binary_children_column.getOffsets();
+        size_t old_chars_size = data.size();
+        size_t old_offsets_size = offsets.size();
+        try
+        {
+            data.insert(reinterpret_cast<const char *>(binary.buf), reinterpret_cast<const char *>(binary.buf) + binary.size);
+            data.push_back(0);
+            offsets.push_back(data.size());
+            static_cast<ColumnUInt8 *>(column_struct->getOrCreateMarkColumn())->insertValue(UInt8(JSONBDataMark::BinaryJSON));
+        }
+        catch (...)
+        {
+            offsets.resize_assume_reserved(old_offsets_size);
+            data.resize_assume_reserved(old_chars_size);
+            throw;
+        }
+    }
+}
+
+struct JSONBinaryConverter
```

Comment:
`simdjson` doesn't support this [RapidJSON handler concept](http://rapidjson.org/classrapidjson_1_1_handler.html). Though this handler concept looks good and I think it can be implemented it in simdjson. I'll try to implement it and suggest a patch to the author of `simdjson`.

Reply:
That's great :+1: